run test.sh with DATA RACE

运行test.sh的时候，遇到data race
queueStart: 0
==================
WARNING: DATA RACE
Read at 0x00c000496408 by goroutine 52:
  sync/atomic.LoadInt64()
      /usr/local/go/src/runtime/race_amd64.s:211 +0xb
  github.com/youzan/nsq/nsqd.(*Topic).flushData()
      /root/nsq/src/github.com/youzan/nsq/nsqd/topic.go:1294 +0x8f
  github.com/youzan/nsq/nsqd.(*Topic).ForceFlush()
      /root/nsq/src/github.com/youzan/nsq/nsqd/topic.go:1262 +0xae
  github.com/youzan/nsq/nsqd.(*NSQD).flushAll()
      /root/nsq/src/github.com/youzan/nsq/nsqd/nsqd.go:715 +0x141
  github.com/youzan/nsq/nsqd.(*NSQD).queueScanLoop()
      /root/nsq/src/github.com/youzan/nsq/nsqd/nsqd.go:875 +0x71c
  github.com/youzan/nsq/nsqd.(*NSQD).Start.func1()
      /root/nsq/src/github.com/youzan/nsq/nsqd/nsqd.go:272 +0x63
  github.com/youzan/nsq/internal/util.(*WaitGroupWrapper).Wrap.func1()
      /root/nsq/src/github.com/youzan/nsq/internal/util/wait_group_wrapper.go:15 +0x5c

Previous write at 0x00c000496408 by goroutine 49:
  [failed to restore the stack]

Goroutine 52 (running) created at:
  github.com/youzan/nsq/internal/util.(*WaitGroupWrapper).Wrap()
      /root/nsq/src/github.com/youzan/nsq/internal/util/wait_group_wrapper.go:13 +0x6f
  github.com/youzan/nsq/nsqd.(*NSQD).Start()
      /root/nsq/src/github.com/youzan/nsq/nsqd/nsqd.go:272 +0xc1
  github.com/youzan/nsq/nsqd.mustStartNSQD()
      /root/nsq/src/github.com/youzan/nsq/nsqd/nsqd_test.go:93 +0x15f
  github.com/youzan/nsq/nsqd.TestTopicResetWithQueueStart()
      /root/nsq/src/github.com/youzan/nsq/nsqd/topic_test.go:549 +0x196
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Goroutine 49 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:388 +0x332
==================
--- FAIL: TestTopicResetWithQueueStart (3.05s)
    testing.go:771: race detected during execution of test
FAIL
coverage: 53.8% of statements
FAIL	github.com/youzan/nsq/nsqd	233.924s

--- a/nsqd/topic_test.go
+++ b/nsqd/topic_test.go
@@ -3,6 +3,8 @@ package nsqd
 import (
 	"errors"
 	"os"
+	"sync/atomic"
+
 	//"runtime"
 	"path"
 	"path/filepath"
@@ -235,6 +237,11 @@ func TestTopicBackendMaxMsgSize(t *testing.T) {
 	test.Equal(t, topic.backend.maxMsgSize, int32(opts.MaxMsgSize+minValidMsgLength))
 }
 
+func changeDynamicConfAutCommit(dynamicConf *TopicDynamicConf) {
+	atomic.StoreInt32(&dynamicConf.AutoCommit, 1)
+	atomic.StoreInt64(&dynamicConf.SyncEvery, 10)
+}
+
 func TestTopicPutChannelWait(t *testing.T) {
 	opts := NewOptions()
 	opts.Logger = newTestLogger(t)
@@ -243,8 +250,7 @@ func TestTopicPutChannelWait(t *testing.T) {
 	defer nsqd.Exit()
 
 	topic := nsqd.GetTopic("test", 0, false)
-	topic.dynamicConf.AutoCommit = 1
-	topic.dynamicConf.SyncEvery = 10
+	changeDynamicConfAutCommit(topic.dynamicConf)
 
 	channel := topic.GetChannel("ch")
 	test.NotNil(t, channel)
@@ -304,9 +310,8 @@ func TestTopicCleanOldDataByRetentionSize(t *testing.T) {
 	defer nsqd.Exit()
 
 	topic := nsqd.GetTopic("test", 0, false)
-	topic.dynamicConf.AutoCommit = 1
-	topic.dynamicConf.SyncEvery = 10
-	topic.dynamicConf.RetentionDay = 1
+	changeDynamicConfAutCommit(topic.dynamicConf)
+	atomic.StoreInt32(&topic.dynamicConf.RetentionDay, 1)
 
 	msgNum := 5000
 	channel := topic.GetChannel("ch")
@@ -377,8 +382,7 @@ func TestTopicCleanOldDataByRetentionDay(t *testing.T) {
 	defer nsqd.Exit()
 
 	topic := nsqd.GetTopic("test", 0, false)
-	topic.dynamicConf.AutoCommit = 1
-	topic.dynamicConf.SyncEvery = 10
+	changeDynamicConfAutCommit(topic.dynamicConf)
 
 	msgNum := 5000
 	channel := topic.GetChannel("ch")
@@ -461,8 +465,7 @@ func TestTopicCleanOldDataByRetentionDayWithResetStart(t *testing.T) {
 	defer nsqd.Exit()
 
 	topic := nsqd.GetTopic("test", 0, false)
-	topic.dynamicConf.AutoCommit = 1
-	topic.dynamicConf.SyncEvery = 10
+	changeDynamicConfAutCommit(topic.dynamicConf)
 
 	readStart := *(topic.backend.GetQueueReadStart().(*diskQueueEndInfo))
 	test.Equal(t, int64(0), readStart.EndOffset.FileNum)
@@ -486,8 +489,7 @@ func TestTopicCleanOldDataByRetentionDayWithResetStart(t *testing.T) {
 
 	nsqd.CloseExistingTopic("test", 0)
 	topic = nsqd.GetTopic("test", 0, false)
-	topic.dynamicConf.AutoCommit = 1
-	topic.dynamicConf.SyncEvery = 10
+	changeDynamicConfAutCommit(topic.dynamicConf)
 
 	writeEnd2 := topic.backend.GetQueueWriteEnd()
 	readStart2 := topic.backend.GetQueueReadStart().(*diskQueueEndInfo)
@@ -551,8 +553,7 @@ func TestTopicResetWithQueueStart(t *testing.T) {
 	defer nsqd.Exit()
 
 	topic := nsqd.GetTopic("test", 0, false)
-	topic.dynamicConf.AutoCommit = 1
-	topic.dynamicConf.SyncEvery = 10
+	changeDynamicConfAutCommit(topic.dynamicConf)
 
 	msgNum := 5000
 	channel := topic.GetChannel("ch")
@@ -610,8 +611,7 @@ func TestTopicResetWithQueueStart(t *testing.T) {
 	// test reopen
 	nsqd.CloseExistingTopic("test", 0)
 	topic = nsqd.GetTopic("test", 0, false)
-	topic.dynamicConf.AutoCommit = 1
-	topic.dynamicConf.SyncEvery = 10
+	changeDynamicConfAutCommit(topic.dynamicConf)
 	channel = topic.GetChannel("ch")
 	test.NotNil(t, channel)
 	newEnd2 := topic.backend.GetQueueWriteEnd().(*diskQueueEndInfo)
@@ -655,8 +655,7 @@ func TestTopicResetWithQueueStart(t *testing.T) {
 	// test reopen
 	nsqd.CloseExistingTopic("test", 0)
 	topic = nsqd.GetTopic("test", 0, false)
-	topic.dynamicConf.AutoCommit = 1
-	topic.dynamicConf.SyncEvery = 10
+	changeDynamicConfAutCommit(topic.dynamicConf)
 	channel = topic.GetChannel("ch")
 	test.NotNil(t, channel)
 	newEnd2 = topic.backend.GetQueueWriteEnd().(*diskQueueEndInfo)

